zebra-chain/src/transaction/sighash.rs

```diff
@@ -0,0 +1,289 @@
+    NetworkUpgrade,
+};
+use blake2b_simd::Hash;
+use byteorder::{LittleEndian, WriteBytesExt};
+use std::io;
+
+const ZCASH_SIGHASH_PERSONALIZATION_PREFIX: &[u8; 12] = b"ZcashSigHash";
+const ZCASH_PREVOUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashPrevoutHash";
+const ZCASH_SEQUENCE_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSequencHash";
+const ZCASH_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashOutputsHash";
+const ZCASH_JOINSPLITS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashJSplitsHash";
+const ZCASH_SHIELDED_SPENDS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSSpendsHash";
+const ZCASH_SHIELDED_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZcashSOutputHash";
+
+pub const SIGHASH_ALL: u32 = 1;
+const SIGHASH_NONE: u32 = 2;
+const SIGHASH_SINGLE: u32 = 3;
+const SIGHASH_MASK: u32 = 0x1f;
+const SIGHASH_ANYONECANPAY: u32 = 0x80;
+
+pub struct SigHasher<'a> {
+    pub(crate) trans: &'a Transaction,
+    pub(crate) hash_type: u32,
+    pub(crate) network: Network,
+    pub(crate) height: BlockHeight,
+}
+
+impl<'a> SigHasher<'a> {
+    pub fn sighash(self) -> Hash {
+        use NetworkUpgrade::*;
+        let mut hash = blake2b_simd::Params::new()
+            .hash_length(32)
+            .personal(&self.personal())
+            .to_state();
+
+        match self.network_upgrade() {
+            Genesis => unimplemented!(),
+            BeforeOverwinter => unimplemented!(),
+            Overwinter | Sapling => self
+                .hash_sighash_zip143(&mut hash)
+                .expect("serialization into hasher never fails"),
+            Blossom => unimplemented!(),
+            Heartwood => unimplemented!(),
+            Canopy => unimplemented!(),
+        }
+
+        hash.finalize()
+    }
+
+    fn network_upgrade(&self) -> NetworkUpgrade {
+        NetworkUpgrade::current(self.network, self.height)
+    }
+
+    fn consensus_branch_id(&self) -> ConsensusBranchId {
+        self.network_upgrade().branch_id().unwrap()
+    }
+
+    /// Sighash implementation for the overwinter and sapling consensus branches
+    fn hash_sighash_zip143<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        self.hash_header(&mut writer)?;
+        self.hash_groupid(&mut writer)?;
+        self.hash_prevouts(&mut writer)?;
+        self.hash_sequence(&mut writer)?;
+        self.hash_outputs(&mut writer)?;
+        self.hash_joinsplits(&mut writer)?;
+        self.hash_lock_time(&mut writer)?;
+        self.hash_expiry_height(&mut writer)?;
+        self.hash_hash_type(&mut writer)?;
+
+        // if let Some((n, script_code, amount)) = transparent_input {
+        //     let mut data = vec![];
+        //     tx.vin[n].prevout.write(&mut data).unwrap();
+        //     script_code.write(&mut data).unwrap();
+        //     data.extend_from_slice(&amount.to_i64_le_bytes());
+        //     (&mut data)
+        //         .write_u32::<LittleEndian>(tx.vin[n].sequence)
+        //         .unwrap();
+        //     h.update(&data);
+        // }
```

Comment:
I'm a bit confused by this commented-out code.

Should it be deleted, or is is blocked on something?

Reply:
this is the upstream code from librustzcash that im porting to work with our data structures, the commented out bit is the part that handles hashing the inputs and is one of the last things I still need to work on from zip-143